ezmlm-limit.1
.TH ezmlm-limit 1
.SH NAME
ezmlm-limit \- Limits traffic to list
.SH SYNOPSIS
.B ezmlm-limit
[
.B -f\fI file
][
.B -dDF
][
.B \-t
.I secs
][
.B -n
.I msgs
]
.I dir

.B ezmlm-limit
is intended to limit traffic to a list in case an outside
source sends excessive messages. Traffic is limited by converting the list
to a message moderated list, or by deferring messages.

.B ezmlm-limit
reads
.IR dir\fB/loopnum .
If it does not exist or contains a time stamp more than
.I secs
seconds old, it is updated.
If it exists, contains a time stamp less than
.I secs
seconds old, and
.I msgs
messages or more have passed since
.I dir\fB/loopnum
was created,
.B ezmlm-limit
creates
.IR dir\fB/modpost .
For lists set up with
.B ezmlm-store(1)
this results in future messages being moderated.

As long as
.I dir\fB/modpost
does not exist,
.B ezmlm-store(1)
will simply forward the message to
.BR ezmlm-send(1) .
Once
.I dir\fB/modpost
exists, messages will be queued for moderation. Thus, in case of error (such
as misconfigured auto responders) subscribers are spared
messages in excess of the limit. However, no messages are lost, since the
moderator(s) can (selectively) approve queued messages. 

If
.I dir\fB/modpost
already exists,
.B ezmlm-limit
will not take any action.

Resetting the list requires manual
removal of
.IR dir\fB/modpost .
Queued messages can be managed with regular moderation
or approved in bulk with
.BR ezmlm-accept(1) .

.B ezmlm-limit
is not required for normal lists, since qmail and
ezmlm loop detection prevent problems caused by regular bounces. However,
it affords some protection against malice, and severely misconfigured
subscribers.

.SH OPTIONS
.TP
.B \-d
Defer excess traffic. Excess messages are deferred and qmail will re-deliver. As
a consequence, traffic will be throttled to the maximum allowed. Due to qmail
delivery back-off, messages may not be disseminated in the order received. If
traffic is continuously above or close to the limit, some messages may be
delayed for a long time or even bounced.
.TP
.B \-D
(Default.)
Make the list moderated once excess traffic is detected. This requires the
list to be configured up with
.B ezmlm-store(1)
rather than
.BR ezmlm-send(1) ,
which is easiest done by first creating a message moderated list, then removing
.IR dir\fB/modpost .
.TP
.B \-n\fI msgs
Trigger after
.I msgs
messages (default 30).
.TP
.B \-f\fI file
Instead of
.I dir\fB/loopnum
use
.I file
as the time stamp. This allows the use of multiple instances of
.B ezmlm-limit
for the same list. For instance, you may allow 60 messages per hour with one
instance and 150 messages per day with another. If
.I file
is relative, this is relative to
.IR dir .
.TP
.B \-F
(Default.)
Use
.I dir\fB/loopnum
as the time stamp.
.TP
.B \-t\fI secs
Check the number of messages within period of
.I secs
seconds (default 3600 seconds).
.SH "SEE ALSO"
ezmlm-accept(1),
ezmlm-make(1),
ezmlm-send(1),
ezmlm-store(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.

- Added support for decoding sender addresses mangled by SRS.

- Fixed bug when running ezmlm-archive on a newly-created list.

- Always enable ezmlm-request in manager.

- Always enable the "deny" blacklist.

- Fixed a long standing bug in ezmlm-store that caused both posting and
  moderating to happen when both modpost and confirmpost were disabled.

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.  Also, the
messages changes also replaced the confirmpost subject line hack in
ezmlm-store and the mailinglist file.  The contents of DIR/confirmpost
and DIR/mailinglist are now ignored in favor of the SUB_CONFIRM_POST and
TXT_MAILING_LIST messages, respectively.

ezmlm-idx-6.0.1, 2007-10-06
===========================


ezmlm-limit.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "stralloc.h"
#include "substdio.h"
#include "readwrite.h"
#include "strerr.h"
#include "sig.h"
#include "lock.h"
#include "getconf.h"
#include "fmt.h"
#include "now.h"
#include "sgetopt.h"
#include "error.h"
#include "scan.h"
#include "open.h"
#include "messages.h"
#include "die.h"
#include "config.h"
#include "idx.h"

const char FATAL[] = "ezmlm-limit: fatal: ";
const char INFO[] = "ezmlm-limit: info: ";
const char USAGE[] =
"ezmlm-limit: usage: ezmlm-limit [-f file] [-dDF] [-n messages] [-t secs] dir";

unsigned long deltasecs = LIMSECS;	/* interval to test over (seconds) */
unsigned long deltanum = LIMMSG;	/* max no messages in interval */
					/* see idx.h. Usually 30 msg/3600 secs*/
int flagd = 0;				/* =0 create modpost, =1 ignore */
					/* excess, =2 defer excess */
int flagloop;
const char *fn = TXT_LOOPNUM;

void die_new(void) { strerr_die2sys(111,FATAL,MSG1(ERR_WRITE,fn)); }

stralloc line = {0};

substdio ssnew;
char newbuf[16];

char strnum[FMT_ULONG];

void main(int argc,char **argv)
{
  char *dir;
  int opt;
  unsigned int pos;
  unsigned long num, loopnum, when;
  unsigned long loopwhen = 0L;
  int fd,fdlock;

  (void) umask(022);
  sig_pipeignore();
  when = (unsigned long) now();

  while ((opt = getopt(argc,argv,"dDf:Fn:t:")) != opteof)
    switch(opt) {
      case 'd': flagd = 1; break;
      case 'D': flagd = 0; break;
      case 'f': if (optarg && *optarg) fn = optarg; break;
      case 'F': fn = TXT_LOOPNUM;
      case 'n':
                if (optarg)
                  scan_ulong(optarg,&deltanum);
                break;
      case 't':
                if (optarg)
                  scan_ulong(optarg,&deltasecs);
                break;
      default:
	die_usage();
  }

  startup(dir = argv[optind++]);

  if (argv[optind])
    die_usage();	/* avoid common error of putting options after dir */
  if (getconf_isset("modpost"))
    _exit(0);		/* already mod */
			/* lock for num and for writing loopnum */
  fdlock = lockfile("lock");

  if (!getconf_ulong(&num,"num",1))
    _exit(99);						/* no msgs */
  if ((flagloop = getconf_line(&line,fn,0))) {
    if(!stralloc_0(&line)) die_nomem();
    pos = scan_ulong(line.s,&loopnum);			/* msg when written */
    if (line.s[pos] == ':')
      scan_ulong(line.s+pos+1,&loopwhen);		/* time written */
  }
  if (!flagloop || loopwhen + deltasecs < when || loopwhen > when) {
					/* loopnum too old, bad or not there */
      fd = open_trunc(fn);		/* no need to write crash-proof */
      if (fd == -1) die_new();
      substdio_fdbuf(&ssnew,write,fd,newbuf,sizeof(newbuf));
      if (substdio_put(&ssnew,strnum,fmt_ulong(strnum,num)) == -1) die_new();
      if (substdio_puts(&ssnew,":") == -1) die_new();
      if (substdio_put(&ssnew,strnum,fmt_ulong(strnum,when)) == -1) die_new();
      if (substdio_puts(&ssnew,"\n") == -1) die_new();
      if (substdio_flush(&ssnew) == -1) die_new();
      close(fd);
  } else if (num >= loopnum + deltanum) {	/* excess messages */
    if (!flagd) {
      if ((fd = open_append("modpost")) == -1)	/* create dir/modpost */
	  strerr_die2sys(111,FATAL,MSG1(ERR_WRITE,"modpost"));
      else {
        close(fd);
        unlink(fn);
        strerr_die2x(0,INFO,MSG(ERR_EXCESS_MOD));
      }
    } else
        strerr_die2x(111,FATAL,MSG(ERR_EXCESS_DEFER));
  }
  _exit(0);
}
